cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,528 @@
+        """
+            Compute all solutions and optionally display the solutions.
+
+            This is the generic implementation, solvers can overwrite this with
+            a more efficient native implementation
+
+            Arguments:
+                - display: either a list of CPMpy expressions, OR a callback function, called with the variables after value-mapping
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - call_from_model: whether the method is called from a CPMpy Model instance or not
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+        if self.objective_given:
+            raise NotImplementedError("Exact does not yet support finding all *optimal* solutions.")
+
+        if not self.solver_is_initialized:
+            self.xct_solver.init([],[])
+            self.solver_is_initialized=True
+            self.xct_solver.setOption("verbosity","0")
+        self.xct_solver.clearAssumptions()
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            self.xct_solver.setOption(kw,str(val))
+
+        solsfound = 0
+        while solution_limit == None or solsfound < solution_limit:
+            # call the solver, with parameters
+            my_status = self.xct_solver.runFull(False,time_limit if time_limit is not None else 0)
+            assert my_status in [0,1,2,3], "Unexpected status code for Exact."
+            if my_status == 0: # found unsatisfiability
+                break
+            elif my_status == 1: # found solution, but not optimality proven
+                assert self.xct_solver.hasSolution()
+                solsfound += 1
+                self.xct_solver.invalidateLastSol() # TODO: pass user vars to this function
+                if display is not None:
+                    self._fillObjAndVars()
+                    if isinstance(display, Expression):
+                        print(display.value())
+                    elif isinstance(display, list):
+                        print([v.value() for v in display])
+                    else:
+                        display()  # callback
+            elif my_status == 2: # found inconsistency
+                assert False, "Error: inconsistency during solveAll should not happen, please warn the developers of this bug"
+            elif my_status == 3: # found timeout
+                return solsfound
+
+        return solsfound
+
+    def has_objective(self):
+        """
+            Returns whether the solver has an objective function or not.
+        """
+        return self.objective_given
+
+
+    def solver_var(self, cpm_var, encoding="onehot"):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
```

Comment:
Maybe also document the non-default `encoding` parameter?

Reply:
I've moved the encoding to a class attribute for Exact, and made "log" default. And checked that this actually works :)

Later, I'd like to make this attribute changeable from the interface, but there was no obvious way now. Let's discuss this later.